src/main/java/edu/kit/compiler/optimizations/ArithmeticIdentitiesOptimization.java

```diff
@@ -0,0 +1,443 @@
+         * of addition and subtraction. `nestedNode` and `constNode` must be
+         * predecessors of `node`, where `constNode` must be a Const. If
+         * `nestedNode` has a Const predecessor, the given operator is used to
+         * fold the Tarvals of the two Consts.
+         */
+        private void foldAdditive(Node node, Node nestedNode, Node constNode,
+                BinaryOperator<TargetValue> accOperator) {
+            assert isAdditiveOperator(node);
+            assert constNode.getOpCode() == ir_opcode.iro_Const;
+
+            var restNode = getRestNode(nestedNode);
+            var leftConst = getConstOperand(nestedNode);
+            var rightConst = getConstValue(constNode);
+
+            if (restNode != null && leftConst != null && isAdditiveOperator(nestedNode)) {
+                // nestedNode is an additive operation with a Const predecessor
+                var accConst = accOperator.apply(leftConst, rightConst);
+                var newConst = graph.newConst(accConst);
+                var newNode = node.getOpCode() != nestedNode.getOpCode()
+                        ? graph.newSub(node.getBlock(), newConst, restNode)
+                        : graph.newAdd(node.getBlock(), restNode, newConst);
+
+                exchange(node, enqueued(newNode));
+            }
+        }
+
+        /**
+         * If possible, simplify the given Mul node using the associative
+         * property of multiplication. The right predecessor of the node
+         * must be a Const.
+         */
+        private void foldMultiplication(Mul node) {
+            assert node.getRight().getOpCode() == ir_opcode.iro_Const;
+
+            var restNode = getRestNode(node.getLeft());
+            var leftConst = getConstOperand(node.getLeft());
+            var rightConst = getConstValue(node.getRight());
+
+            if (restNode != null && leftConst != null
+                    && node.getLeft().getOpCode() == ir_opcode.iro_Mul) {
+                var constNode = graph.newConst(leftConst.mul(rightConst));
+                var newNode = graph.newMul(node.getBlock(), restNode, constNode);
+
+                exchange(node, enqueued(newNode));
+            }
+        }
+
+        /**
+         * Replace the given Div or Mode node. Result projections are replaced
+         * with `val`. Memory projections are replaced with `mem`.
+         */
+        private void exchangeDivOrMod(Node node, Node mem, Node val) {
+            assert node.getOpCode() == ir_opcode.iro_Div
+                    || node.getOpCode() == ir_opcode.iro_Mod;
+
+            for (var edge : BackEdges.getOuts(node)) {
+                if (edge.node.getMode().equals(Mode.getM())) {
+                    exchange(edge.node, mem);
+                } else if (edge.node.getMode().equals(val.getMode())) {
+                    exchange(edge.node, val);
+                } else {
+                    throw new UnsupportedOperationException(
```

Comment:
Wouldn't it be enough to simply not exchange the node?

Reply:
Yes, however if we were to start using the control flow projections of `Div` and `Mod`, this code might break in a less predictable way. I'm just being a bit defensive here.